src/psyclone/psyir/nodes/routine.py

```diff
@@ -64,19 +73,29 @@ class Routine(Schedule, CommentableMixin):
     _children_valid_format = "[Statement]*"
     _text_name = "Routine"
 
-    def __init__(self, name, is_program=False, **kwargs):
+    def __init__(self, name, is_program=False, symbol=None,
+                 symbol_tag=None, **kwargs):
+        # These attributes need to be set before anything, as the _symbol
+        # is required for setting the _parent, which is overriden by Routine
+        self._symbol = symbol
+        self._symbol_tag = symbol_tag
+        self._symbol_in_table = False
+        self._parent_node = None
         super().__init__(**kwargs)
 
         self._return_symbol = None
-        self._name = None
-        # Name is set-up by the name setter property
-        self.name = name
-
         if not isinstance(is_program, bool):
             raise TypeError(f"Routine argument 'is_program' must be a bool "
                             f"but got '{type(is_program).__name__}'")
         self._is_program = is_program
 
+        # Symbol is set-up by the name setter property.
+        # name property is taken from the symbol.
+        # We do this last to avoid any other errors resulting in a
+        # symbol in a SymbolTable that we shouldn't have.
+        self._has_set_name = False
+        self.name = name
```

Comment:
It is unclear to me what takes precedence, the name or the symbol.name? But regardless, I think it would be much simpler if we could just have one (and fix #1200) so there is no conflict.

I guess both cases are useful sometime, but maybe we could use the fact that we have 2 constructors (init and create) and have them use a different argument.

Usually use try to have the init leaner and we use it for top-down creation, so I suggest removing the "name" here and make the symbol mandatory. But then, we usually have logic to create inner object in the "create" method and use use it for bottom-up tree creation (which means the parent symbol table may even not exist). So I suggest removing symbol from there and make the method always create the RoutineSymbol for the given name. 

This would require adding/deleting the .create from current uses (including specialisations e.g.[GO/Dyn]KernelSchedule) but I believe it will make the implementation easier and the API more predictable. What do you think?

Reply:
I'm a bit confused on this comment - the routine has no `_name`, the `name` property and setter are entirely based upon the symbol object creation or passed in.

The setup in Routine initialisation are as follows:
1. There was no `symbol` argument, so we create a `RoutineSymbol` with the `name` argument's name.
2. We take the symbol (either created or passed as an argument) and add it to the parent's symbol table
3. We set `self._has_set_name = True` so future calls to `name` will take the alternative code path where the symbol's name is update.

The one thing I should probably do is check that the `symbol.name` and `name` are the same if `symbol` is provided, or I guess **make name optional and require either that `name` or `symbol` is provided which I think is probably best?** @sergisiso 

This is essentially doing what you want, but it does it all from inside the `name` setter instead of it relying on modifying other sections of the code (and means we don't need to modify the rest of the code). Its a little opaque perhaps but I don't think it should affect any subclasses unless they override `name`.

Reply:
To clarify, my comment was only about the duplication in the input arguments: "name" and "symbol".name. As you say one solution would be to "require either that name or symbol", but the other solution is to provide just symbol in the init (and discard all the _has_set_name). And then move the string "name" option to the "create" method (and remove symbol from there), in this case the create method would first create the RoutineSymbol with the given name, and then call init with it.

This would match the style of other node types where we keep the init as lean as possible, but we have create arguments that accept strings and get transformed to the underlying needed nodes with the logic inside create.

What I don't know is if this will help simplify the current _has_set_name/ name setter implementation. If possible I would prefer not to have a "routine.name setter", and rename routines by renaming their symbol with the symbol_table.rename_symbol method explicitly, as this is the same we need to do for other symbols.

(I am aware that this will require changes in the current routine usage code, because some `Routine("name")` will need to be updated to `Routine.create("name")` and `routine.name = "new_name"` will need to be updated to `routine.scope.symbol_table.rename_symbol(routine.symbol, "new_name")` )

Would this help make the implementation easier, or are there problems that I am not seeing?

Reply:
btw, if they are not attached to an scope (as you mention in another comment), could we do:
`routine.symbol = RoutineSymbol("new_name")`?
which is the same API that we have for References

Reply:
> I'm happy to require Routine to require the symbol input argument and change the name to be in create, that seems ok.

This will have a lot of side effects in its subclasses and probably make the create function less clean (it will need kwargs since subclasses have extra subclasses) but i think thats ok.

Reply:
Before you do the work, what are the kwargs/subclasses that we will need?